src/pubtools/sign/signers/cosignsigner.py

```diff
@@ -0,0 +1,291 @@
+        },
+        default="https://rekor.sigstore.dev",
+    )
+    env_variables: Dict[str, str] = field(
+        init=False,
+        metadata={
+            "description": "environment variables used for signing",
+            "sample": '{"AWS_ACCESS_KEY_ID": "xxxxxxx", "AWS_SECRET_ACCESS_KEY":"yyyyyyyyy"}',
+        },
+        default_factory=dict,
+    )
+    upload_tlog: bool = field(
+        init=False,
+        metadata={"description": "upload signing record to rekor", "sample": "False"},
+        default=True,
+    )
+    log_level: str = field(
+        init=False, metadata={"description": "Log level", "sample": "debug"}, default="info"
+    )
+
+    SUPPORTED_OPERATIONS: ClassVar[List[SignOperation]] = [
+        ContainerSignOperation,
+    ]
+
+    _signer_config_key: str = "cosign_signer"
+
+    def load_config(self: CosignSigner, config_data: Dict[str, Any]) -> None:
+        """Load configuration of messaging signer."""
+        self.cosign_bin = config_data["cosign_signer"].get("cosign_bin", self.cosign_bin)
+        self.timeout = config_data["cosign_signer"].get("timeout", self.timeout)
+        self.allow_http_registry = config_data["cosign_signer"].get(
+            "allow_http_registry", self.allow_http_registry
+        )
+        self.allow_insecure_registry = config_data["cosign_signer"].get(
+            "allow_insecure_registry", self.allow_insecure_registry
+        )
+        self.rekor_url = config_data["cosign_signer"].get("rekor_url", self.rekor_url)
+        self.upload_tlog = config_data["cosign_signer"].get("upload_tlog", self.upload_tlog)
+        self.env_variables = config_data["cosign_signer"].get("env_variables", self.env_variables)
+
+    def operations(self: CosignSigner) -> List[SignOperation]:
+        """Return list of supported operations."""
+        return self.SUPPORTED_OPERATIONS
+
+    def sign(self: CosignSigner, operation: SignOperation) -> SigningResults:
+        """Run signing operation.
+
+        :param operation: signing operation
+        :type operation: SignOperation
+
+        :return: SigningResults
+        """
+        if isinstance(operation, ContainerSignOperation):
+            return self.container_sign(operation)
+        else:
+            raise UnsupportedOperation(operation)
+
+    def container_sign(self: CosignSigner, operation: ContainerSignOperation):
+        """Run container signing operation.
+
+        :param operation: signing operation
+        :type operation: ContainerSignOperation
+
+        :return: SigningResults
+        """
+        set_log_level(LOG, self.log_level)
+        if operation.references and len(operation.digests) != len(operation.references):
+            raise ValueError("Digests must pairs with references")
+
+        signer_results = CosignSignerResults(status="ok", error_message="")
+
+        operation_result = ContainerSignResult(signing_key=operation.signing_key, signed_claims=[])
+
+        signing_results = SigningResults(
```

Comment:
Reason for that is you can call dataclasses.asdict on it and it will generate you nice structured dictionary report of signing result. And you will exactly know what is what. If you allow multiple SigningResults it won't be possible to force users to keep same structure. Now you know operation and operation_results will be same and only signer and signer_results can be different based on what type of signer you use